Anomaly: new  binary output mode

breznak opened the pull request:
Fixes #1840

rhyolight commented:
@subutai or @scottpurdy ?

scottpurdy commented:
@breznak - I will take a look at this tomorrow. One initial thing is that it would be nice if you squashed the initial commits together. It is ok to have more than one if there are logical units of work done but we should try to keep the commit history somewhat clean.

breznak commented:
:+1: thanks! Squashed (maybe too much :package: )

scottpurdy commented on nupic/algorithms/anomaly.py:
```diff
@@ -140,6 +148,17 @@ def compute(self, activeColumns, predictedColumns,
+        score = 0.0
+
+    # final check on score within [0,1]
+    score = int(score * 100000000)/float(100000000) # 7 dec places
```
you shouldn't have to do this

breznak replied:
i do, probably due to numeric instability bug (in numpy?) - I was getting a CLA run with anomaly scores = x.xxxx*10e-13

scottpurdy replied:
as long as it is positive it shouldn't be a problem?

breznak replied:
sorry, my bad. but the problem was sum(anomalyScore) was negative... should I remove the rounding and let assert prove/catch it?

scottpurdy commented on nupic/algorithms/anomaly.py:
```diff
@@ -140,6 +148,17 @@ def compute(self, activeColumns, predictedColumns,
+
+    # final check on score within [0,1]
+    score = int(score * 100000000)/float(100000000) # 7 dec places
+    assert score >= 0 and score <= 1
```
`assert score >= 0.0 and score <= 1.0`

scottpurdy commented on nupic/algorithms/anomaly.py:
```diff
@@ -100,6 +102,12 @@ def __init__(self, slidingWindowSize = None, mode=MODE_PURE):
                        "Anomaly.MODE_WEIGHTED; you used: %r" % self._mode)
+    self._binaryThreshold = binaryAnomalyThreshold
+    if binaryAnomalyThreshold is not None and (
+          (binaryAnomalyThreshold >= 1  or binaryAnomalyThreshold <= 0) or
```
`(binaryAnomalyThreshold >= 1.0  or binaryAnomalyThreshold <= 0.0)`

scottpurdy commented on nupic/algorithms/anomaly.py:
```diff
@@ -100,6 +102,12 @@ def __init__(self, slidingWindowSize = None, mode=MODE_PURE):
+    self._binaryThreshold = binaryAnomalyThreshold
+    if binaryAnomalyThreshold is not None and (
+          (binaryAnomalyThreshold >= 1  or binaryAnomalyThreshold <= 0) or
+          not isinstance(binaryAnomalyThreshold, float) ):
```
put check for whether it is a float before the bounds checks

scottpurdy commented on nupic/algorithms/anomaly.py:
```diff
@@ -140,6 +148,17 @@ def compute(self, activeColumns, predictedColumns,
       score = self._movingAverage.next(score)
 
+    # apply binary discretization if required
+    if self._binaryThreshold is not None:
```
In the case where there is a binary threshold specified and it is also doing a moving average, should the threshold be applied before or after the moving average? Make this clear in the doc string for the binary threshold in the constructor.

breznak replied:
good point! actually, both make some sence. What do you think? 
CC @subutai

scottpurdy replied:
Are you adding this for a real need that you have or just because you think it may be useful to someone? In the former case, which is useful for you?

breznak replied:
I am using `moving avg, then discretize` to have fewer but compact packets of anomaly detected events; now I don't see much utility in doing it the other way. 

Actualy, the problem of ordering when to do MA gives a lot of possibilities we could judge. 

`pure _ (opt. likelihood/weighted) _ (discretize) _` where `moving average` can be done at any `_`

subutai replied:
We always apply a threshold after the likelihood calculation. Thresholding the raw anomaly score (whether averaged or not) usually doesn't work.

scottpurdy commented:
You will need to set the binary threshold attribute in the `__setstate__` method or old models will end up with exceptions.

breznak commented:
> You will need to set the binary threshold attribute in the __setstate__ method or old models will end up with exceptions.

I do in https://github.com/numenta/nupic/pull/1845/files#diff-5bf13635ee2cded8c4bfe689ea9a297dR194 

..thank you for review @scottpurdy , I addressed the issues

scottpurdy commented on nupic/algorithms/anomaly.py:
```diff
@@ -100,6 +103,13 @@ def __init__(self, slidingWindowSize = None, mode=MODE_PURE):
+    self._binaryThreshold = binaryAnomalyThreshold
+    if binaryAnomalyThreshold is not None and ( 
+          not isinstance(binaryAnomalyThreshold, float) or
+          binaryAnomalyThreshold >= 1.0  or 
```
actual since we are checking out of bounds this and next line should be strictly `>` and `<`

breznak replied:
actually, I defined the threshold on open interval (0,1) because 0/1 values would not make much sense: no anomalies/all anomalies.

scottpurdy commented:
With regards to backwards compatibility, the code you linked to just shows that it will be set for new instances. Old serialized models will not have it set and will break when you try to use them. If you aren't familiar with Python serialization with pickle, please read up on pickle, `__getstate__`, and `__setstate__`.

breznak commented:
done

breznak commented:
@scottpurdy this should be ready..

scottpurdy commented on nupic/utils.py:
```diff
@@ -83,3 +83,12 @@ def next(self, newValue):
+
+    if not hasattr(self, "total"):
+      self.total = 0
+      self.slidingWindow = []
```
If the serialized version had a sliding window do you really want to reset it? How about replace these lines with:

```
self.total = sum(self.slidingWindow)
```

breznak commented:
@scottpurdy thanks! fixed. ready?

rhyolight commented:
@scottpurdy please approve.
@breznak please merge in master.

scottpurdy commented:
@breznak - Do you get better results with this functionality than with anomaly likelihood? To @subutai 's point, we did a lot of work figuring out how to compute useful anomaly scores and this seems contrary to our experience. I'm ok merging this, but want to make sure it is actually useful code.

breznak commented:
@scottpurdy I actually do. 
* This is more like a post-processing filter, so it's combinable with the likelihood mode. 
* it replaces existing functionality of `autoDetectThreshold`, just moving it from ClassifierRegion to Anomaly where it logically belongs
* my aim for this is using it in AnomalyMetric to reduce anomalies to certain percentage.

rhyolight commented:
@breznak Are there any docs?

breznak commented:
Not but the code comments/param descriptions. Should there be more?

rhyolight commented:
I see. I think that's good for now. The code docs will update themselves with the new param.
